components/engine_rocks/src/misc.rs

```diff
@@ -1,90 +1,187 @@
-        let handle = util::get_cf_handle(self.as_inner(), cf)?;
-        Ok(self.as_inner().flush_cf(handle, sync)?)
+impl RocksEngine {
+    fn is_titan(&self) -> bool {
+        self.as_inner().is_titan()
     }
 
-    fn delete_files_in_range_cf(
+    // We store all data which would be deleted in memory at first because the data of region will never be larger than
+    // max-region-size.
+    fn delete_all_in_range_cf_by_ingest(
         &self,
         cf: &str,
-        start_key: &[u8],
-        end_key: &[u8],
-        include_end: bool,
+        sst_path: String,
+        ranges: &[Range],
     ) -> Result<()> {
-        let handle = util::get_cf_handle(self.as_inner(), cf)?;
-        Ok(self
-            .as_inner()
-            .delete_files_in_range_cf(handle, start_key, end_key, include_end)?)
-    }
-
-    fn delete_blob_files_in_range_cf(
-        &self,
-        cf: &str,
-        start_key: &[u8],
-        end_key: &[u8],
-        include_end: bool,
-    ) -> Result<()> {
-        if self.as_inner().is_titan() {
-            let handle = util::get_cf_handle(self.as_inner(), cf)?;
-            self.as_inner().delete_blob_files_in_range_cf(
-                handle,
-                start_key,
-                end_key,
-                include_end,
-            )?;
+        let mut ranges = ranges.to_owned();
+        ranges.sort_by(|a, b| a.start_key.cmp(b.end_key));
+        let mut opts = IterOptions::new(None, None, false);
+        if self.is_titan() {
+            // Cause DeleteFilesInRange may expose old blob index keys, setting key only for Titan
+            // to avoid referring to missing blob files.
+            opts.set_key_only(true);
         }
 
-        Ok(())
-    }
-
-    fn delete_all_in_range_cf(
-        &self,
-        cf: &str,
-        start_key: &[u8],
-        end_key: &[u8],
-        use_delete_range: bool,
-    ) -> Result<()> {
-        let mut wb = self.write_batch();
-        if use_delete_range && cf != CF_LOCK {
-            wb.delete_range_cf(cf, start_key, end_key)?;
-        } else {
-            let start = KeyBuilder::from_slice(start_key, 0, 0);
-            let end = KeyBuilder::from_slice(end_key, 0, 0);
-            let mut opts = IterOptions::new(Some(start), Some(end), false);
-            if self.as_inner().is_titan() {
-                // Cause DeleteFilesInRange may expose old blob index keys, setting key only for Titan
-                // to avoid referring to missing blob files.
-                opts.set_key_only(true);
+        let mut writer_wrapper: Option<RocksSstWriter> = None;
+        let mut data: Vec<Vec<u8>> = vec![];
+        let mut last_end_key: Option<Vec<u8>> = None;
+        for r in ranges {
+            // There may be a range overlap with next range
+            if last_end_key
+                .as_ref()
+                .map_or(false, |key| key.as_slice() > r.start_key)
+            {
+                self.delete_all_in_range_cf_by_key(cf, &r)?;
```

Comment:
If there are only a few keys overlapped, we still need to delete all keys in the range? How about calling `next` on the iterator directly?

Reply:
Yes. But what you mean can only do for `cleanup_overlap_regions`. In most case we should delete keys for regions which have been destroye even if they are not overlapped with the region that acquire applying snapshot.